uc-crux-llvm/src/UCCrux/LLVM/Classify.hs

```diff
@@ -0,0 +1,643 @@
+               Maybe (Some (TypedSelector m arch argTypes))
+             getPtrOffsetAnn ptr = getTermAnn (LLVMPtr.llvmPointerOffset ptr)
+             getPtrBlockAnn ::
+               LLVMPtr.LLVMPtr sym w ->
+               IO (Maybe (Some (TypedSelector m arch argTypes)))
+             getPtrBlockAnn ptr =
+               do
+                 int <- What4.natToInteger sym (LLVMPtr.llvmPointerBlock ptr)
+                 pure $ getTermAnn int
+             getAnyPtrOffsetAnn ::
+               LLVMPtr.LLVMPtr sym w ->
+               [Some (TypedSelector m arch argTypes)]
+             getAnyPtrOffsetAnn ptr =
+               let subAnns =
+                     case What4.asApp (LLVMPtr.llvmPointerOffset ptr) of
+                       Nothing -> []
+                       Just app -> foldMapFC (maybeToList . getTermAnn) app
+                in case getPtrOffsetAnn ptr of
+                     Just ann -> ann : subAnns
+                     Nothing -> subAnns
+
+             argName :: Ctx.Index argTypes tp -> String
+             argName idx = funCtx ^. argumentNames . ixF' idx . to getConst . to (maybe "<top>" Text.unpack)
+
+             elemsFromOffset ::
+               What4.ConcreteVal (What4.BaseBVType w) ->
+               PartTypeRepr m ft ->
+               Int
+             elemsFromOffset offset partType =
+               let dl = modCtx ^. dataLayout
+                   pointedTo = asFullType (funCtx ^. moduleTypes) partType
+                   typeSize = bytesToInteger (memTypeSize dl (toMemType pointedTo))
+                in 1 + fromIntegral (asUnsigned (What4.fromConcreteBV offset) `div` fromIntegral typeSize)
+
+             oneArgConstraint :: Ctx.Index argTypes inTy -> Cursor m inTy atTy -> Constraint m atTy -> [NewConstraint m argTypes]
+             oneArgConstraint idx cursor constraint =
+               [NewConstraint (SomeInSelector (SelectArgument idx cursor)) constraint]
+
+             oneArgShapeConstraint :: Ctx.Index argTypes inTy -> Cursor m inTy atTy -> ShapeConstraint m atTy -> [NewConstraint m argTypes]
+             oneArgShapeConstraint idx cursor shapeConstraint =
+               [NewShapeConstraint (SomeInSelector (SelectArgument idx cursor)) shapeConstraint]
+
+             unclass =
+               do
+                 liftIO $ (appCtx ^. log) Hi ("Couldn't classify error." :: Text)
+                 pure $ ExUncertain (UUnclassified (ppBB badBehavior))
+
+             unfixed :: Unfixed -> IO (Explanation m arch argTypes)
+             unfixed tag =
+               do
+                 (appCtx ^. log) Hi "Prognosis: Fixable, but the fix is not yet implemented."
+                 pure $ ExUncertain (UUnfixed tag)
+
+             unfixable :: Unfixable -> IO (Explanation m arch argTypes)
+             unfixable tag =
+               do
+                 (appCtx ^. log) Hi "Prognosis: Don't know how to fix this."
+                 pure $ ExUncertain (UUnfixable tag)
```

Comment:
Just as a general readability comment - I can follow this, but it would be a lot easier to follow if these bindings were extracted to the top level or a where clause. It would make the rest of the structure of this function a lot clearer

Reply:
Maybe this is a bit pedantic, but I'd like to understand the general principle here so I can use to write more readable code in the future :-)

>  I can follow this, but it would be a lot easier to follow if these bindings were extracted to the top level

Here's what I see as the pros of extracting to a top-level binding:

- It's more clear in the body of each utility function what variables are captured from the external environment or arguments, basically because there's less in scope.
- It makes it easier to reuse this code in the future if need be

And the cons:

- It makes it less clear that these utilitiy functions actually are just used inside `classify`
- A few more variables have to be passed as arguments that were previously closed over from the arguments of `classify` (mostly the `annotations` map)

Am I missing anything here? If not, we might just have different personal weightings of the above factors, and I'm more than happy to adjust to use more top-level bindings.

> or a where clause

This one I'm a lot fuzzier on. For small things, I like `let` bindings because you see the definition of a name before its use, which I find easier to follow (though that's a completely unfounded opinion). Since they pretty much capture the same variables and are usable in the same scope, do you have any general guidelines on `let` vs. `where`? [There is some information out there](https://wiki.haskell.org/Let_vs._Where), but I haven't found anything with well-grounded guidelines on the subject.

Reply:
My personal opinions: I start to get uncomfortable if a group of let or where bindings can't all fit on my screen at once, and I start to think seriously about making things top-level.  If I'm giving something a type signature, I'll think even more about making it top-level.  I consider `let` to be for small things (one-liners mostly), `where` for medium sized things and top-level bindings for anything "big", and I'm not shy about promoting definitions through the levels.  Most especially, I don't much like to see nesting (lets inside lets, lets inside where, etc) as the scoping issues really start to get gnarly.

Additional personal opinion while I'm here: not much of a fan of `flip` as a pipefitting technique.  I would much rather see a do-notation binding.

Reply:
> My personal opinions: 

Thanks for sharing that, @robdockins! I'll write that down as a guideline for myself and I'll update this part of the code to match.

Might also be worth a mention on the Galois Haskell Programming Styles Confluence page.

Reply:
For what it is worth, my reasons align closely to Rob's. By the time I scroll through all of the helpers, I've lost the context of what they are helping with. I like to avoid nested lets because I find them confusing. Nested type signatures just give me extra chances to accidentally shadow type variable names in ways I wasn't expecting (or accidentally grab a type variable from an outer scope that I wasn't expecting).

I'm less concerned with future reuse of the helpers, though I've definitely spent more time than I'd like untangling local variable captures when I do need to extract a single helper function for reuse, only to find that it is part of a mutually recursive let group with lots of variable captures.

It is more typing, but I try to optimize for me in 6 months, who can be assumed to not know anything.